iob_soc.py

```diff
@@ -33,325 +34,325 @@
-            },
-            # mandatory parameters (do not change them!)
-            {
-                "name": "ADDR_W",
-                "type": "F",
-                "val": "32",
-                "min": "1",
-                "max": "32",
-                "descr": "Address bus width",
-            },
-            {
-                "name": "DATA_W",
-                "type": "F",
-                "val": "32",
-                "min": "1",
-                "max": "32",
-                "descr": "Data bus width",
-            },
-            {
-                "name": "AXI_ID_W",
-                "type": "F",
-                "val": "0",
-                "min": "1",
-                "max": "32",
-                "descr": "AXI ID bus width",
-            },
-            {
-                "name": "AXI_ADDR_W",
-                "type": "F",
-                "val": "`IOB_SOC_MEM_ADDR_W",
-                "min": "1",
-                "max": "32",
-                "descr": "AXI address bus width",
-            },
-            {
-                "name": "AXI_DATA_W",
-                "type": "F",
-                "val": "`IOB_SOC_DATA_W",
-                "min": "1",
-                "max": "32",
-                "descr": "AXI data bus width",
-            },
-            {
-                "name": "AXI_LEN_W",
-                "type": "F",
-                "val": "4",
-                "min": "1",
-                "max": "4",
-                "descr": "AXI burst length width",
-            },
-            {
-                "name": "MEM_ADDR_OFFSET",
-                "type": "F",
-                "val": "0",
-                "min": "0",
-                "max": "NA",
-                "descr": "Offset of memory address",
-            },
-        ]
-        self.ios = [
-            {
-                "name": "clk_en_rst",
-                "type": "slave",
-                "port_prefix": "",
-                "wire_prefix": "",
-                "descr": "Clock, enable, and reset",
-                "ports": [],
-            },
-            {
-                "name": "trap",
-                "type": "master",
-                "port_prefix": "",
-                "wire_prefix": "",
-                "descr": "iob-soc trap signal",
-                "ports": [
-                    {
-                        "name": "trap",
-                        "direction": "output",
-                        "width": 1,
-                        "descr": "CPU trap signal",
-                    },
-                ],
-            },
-            {
-                "name": "axi",
-                "type": "master",
-                "wire_prefix": "",
-                "port_prefix": "",
-                "mult": "",  # Will be filled automatically
-                "widths": {
-                    "ID_W": "AXI_ID_W",
-                    "ADDR_W": "AXI_ADDR_W",
-                    "DATA_W": "AXI_DATA_W",
-                    "LEN_W": "AXI_LEN_W",
+    version = "V0.70"
```

Comment:
Types are used whenever we want arguments to be automatically checked.

We are not adding types to inflict pain upon ourselves but just to make life easier for users (in case they pass the wrong arguments) and for us who don't have to write meaningless code to check stuff.

Reply:
before we did not have this feature. Now we have. So we might as well use it!

Reply:
It may be time to have all those dicts changed into classes!

Reply:
a class creates a type, which can now be easily checked

Reply:
> It may be time to have all those dicts changed into classes!

I would start by creating the classes one by one so we can discuss them and then apply them and start writing the methods

Reply:
The [block_groups](https://github.com/IObundle/iob-soc/pull/798/files#diff-551ef7ec958375dc7823b60fea7a6b8ff010a80b79c53afdd937d66369e41807R322-R346) list is one of those lists that we already modified to use objects instead of dictionaries.

Reply:
I suggest we simplify it to just blocks, a list of blocks only. 

It is currently used to document top-level blocks. 

The groups were created out of laziness -- so that we don't have to group verilog modules inside a larger verilog module

Ideally, an IP core should have a reasonably low number of top-level blocks (< 10 as a rule of thumb), and only these are documented.

Because we had too many we created block groups to artificially decrease them.

In fact, we will not need blocks , once we create IP cores out of other modules.

My suggestion is to delete block groups and manually write the list in the tex file.

In the future this list is easily extracted from the description of the top-level blocks (blocks below the top-level)

Reply:
Not simplify! Simply erase.

Reply:
Should I delete the `block_groups` list completely, or modify it to be a `blocks` list (without groups)?

> In the future this list is easily extracted from the description of the top-level blocks (blocks below the top-level)

From this phrase it seems that we still have a list describing the blocks. Is this the `blocks` list (without groups)?